Collections/src/Electron.cc

```diff
@@ -235,38 +272,105 @@ void
-    {
-      passesTightID_noIsolation_ = (this->full5x5_sigmaIetaIeta ()                                                        <   0.0101
-                                 && fabs (this->deltaEtaSuperClusterTrackAtVtx ())                                        <   0.00926
-                                 && fabs (this->deltaPhiSuperClusterTrackAtVtx ())                                        <   0.0336
-                                 && this->hadronicOverEm ()                                                               <   0.0597
-                                 && fabs (1.0 / this->ecalEnergy () - this->eSuperClusterOverP () / this->ecalEnergy ())  <   0.012
-                                 && fabs (this->gsfTrack ()->dxy (vertex.position ()))                                    <   0.0111
-                                 && fabs (this->gsfTrack ()->dz (vertex.position ()))                                     <   0.0466
-#if CMSSW_VERSION_CODE >= CMSSW_VERSION(9,4,0)
-                                 && this->gsfTrack ()->hitPattern ().numberOfAllHits (reco::HitPattern::MISSING_INNER_HITS)  <=  2
-#else
-                                 && this->gsfTrack ()->hitPattern ().numberOfHits (reco::HitPattern::MISSING_INNER_HITS)  <=  2
-#endif
-                                 && !ConversionTools::hasMatchedConversion (*this, conversions, beamspot.position ()));
-    }
-  else if (fabs (this->superCluster ()->eta ()) < 2.5)
-    {
-      passesTightID_noIsolation_ = (this->full5x5_sigmaIetaIeta ()                                                        <   0.0279
-                                 && fabs (this->deltaEtaSuperClusterTrackAtVtx ())                                        <   0.00724
-                                 && fabs (this->deltaPhiSuperClusterTrackAtVtx ())                                        <   0.0918
-                                 && this->hadronicOverEm ()                                                               <   0.0615
-                                 && fabs (1.0 / this->ecalEnergy () - this->eSuperClusterOverP () / this->ecalEnergy ())  <   0.00999
-                                 && fabs (this->gsfTrack ()->dxy (vertex.position ()))                                    <   0.0351
-                                 && fabs (this->gsfTrack ()->dz (vertex.position ()))                                     <   0.417
+
 #if CMSSW_VERSION_CODE >= CMSSW_VERSION(9,4,0)
-                                 && this->gsfTrack ()->hitPattern ().numberOfAllHits (reco::HitPattern::MISSING_INNER_HITS)  <=  1
-#else
-                                 && this->gsfTrack ()->hitPattern ().numberOfHits (reco::HitPattern::MISSING_INNER_HITS)  <=  1
-#endif
-                                 && !ConversionTools::hasMatchedConversion (*this, conversions, beamspot.position ()));
+  // https://twiki.cern.ch/twiki/bin/view/CMS/CutBasedElectronIdentificationRun2#Offline_selection_criteria
```

Comment:
I don't understand several things this section:
- Why should we keep anything in Electron.cc that explicitly references all the cuts in a given ID? Isn't the point to remove all things like this? In other words, why define passesTightID_noIsolation_ at all?
- How does it make sense without the &&s?
- Why does this definition of passesTightID_noIsolation_ include dxy and dz cuts, when the electron ID twiki says that they are not included in the ID any more?
- I understand that numberOfLostHits is better for disappearing tracks, but numberOfAllHits is probably what should be used for displaced SUSY. So, what in the end is implemented? I'm confused.

Reply:
- I kept this for the moment for backwards compatibility and to compare the two methods. I agree that the point is to remove it and we should, but I'd like to review this in light of disappearing tracks first. Specifically, since the isolation is baked into the VID, if we need this WITHOUT isolation then I need to find a method to accomplish that.
- The &&s are just moved to the right side of the lines. This is purely a conventional change, I just don't think any other such lines are done with the &&s on the left in this repo.
- This is probably an artifact of Andrew writing it for disappearing tracks and Jamie never using this function for displaced SUSY. Disappearing tracks requires these cuts, and no one complained that they were in it.
- It's not necessarily better for disappearing tracks, I just meant that it's an interesting and easily overlooked change. The VID in 94X uses LostHits, which is outside of our control, and I changed our own function to match EGM's logic. In the end LostHits is a looser cut than AllHits here -- in my comparison test I saw one electron that had a hit pattern of (INVALID, MISSING) which failed AllHits<=1 but passed LostHits<=1 and passed the VID. For displaced SUSY, you will need to ask yourself if such cases are acceptable for your analysis, but we should just highlight that such a thing is happening when the VID is used.